### PR TITLE
Nested logger context + refactoring

### PR DESCRIPTION
**Pull request by nukemberg**
Slightly changed the API of wryte, hence the minor version bump:
- Added nested logger support, see `README.md`
- While logging-time methods should never blow up (wryte core principle?), config-time methods _should_ blow up. You don't want logger config to silently fail and leave you without any logs...

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+[[source]]
```
Comment (nir0s) [inline Pipfile:1]: Thanks for this :)

File: README.md [hunk 2]
```diff
@@ -318,6 +319,22 @@ wryter.unbind('user_id')
 ```
 
 And just like with the logger itself, you can bind nested dicts, kwargs and JSON strings.
+### Nested logger context
+
+In many cases you want to bind some information specific to the context, without modifying the original binded parameters. Wryte supports this using the `nested` method:
+
+```python
+nested_wryter = wryter.nested(name='nested', {'user_id': framework.user}, key=value)
```
Comment (nir0s) [inline README.md:327]: You have a named parameter before a non-named one :)

File: wryte.py [hunk 3]
```diff
@@ -22,6 +22,7 @@
 import logging
 import logging.handlers
 from datetime import datetime
+import copy
```
Comment (nir0s) [inline wryte.py:25]: Nitpicking. We're ordering imports by length. Please move it up,

File: wryte.py [hunk 4]
```diff
@@ -58,17 +59,23 @@
...
-    'debug': logging.DEBUG,
-    'info': logging.INFO,
-    'warning': logging.WARNING,
-    'warn': logging.WARN,
-    'error': logging.ERROR,
-    'critical': logging.CRITICAL,
-    'event': logging.INFO
-}
+if hasattr(logging, '_nameToLevel'):
+    LEVEL_CONVERSION = logging._nameToLevel
```
Comment (nir0s) [inline wryte.py:63]: Wonderful!

Comment (nir0s) [reply wryte.py:63]: Though I'd "Ask for forgiveness, not permission" here. Makes more sense to me than `hasattr`.

**Comment by nir0s** [comment]
Windows build has failed.

File: wryte.py [hunk 5]
```diff
@@ -58,17 +59,23 @@
...
+else: # python 2.x
+    LEVEL_CONVERSION = dict((level_str, level_num) for (level_str, level_num) in logging._levelNames.items() if isinstance(level_str, str))
```
Comment (nir0s) [inline wryte.py:65]: This won't pass linting (max 120 chars)

File: wryte.py [hunk 6]
```diff
@@ -602,44 +609,86 @@ def log(self, level, message, *objects, **kwargs):
...
 
     def error(self, message, *objects, **kwargs):
         if '_set_level' in kwargs:
-            self.set_level(kwargs['_set_level'])
-        obj = self._enrich(message, 'error', objects, kwargs)
-        self.logger.error(obj)
+            try:
+                self.set_level(kwargs['_set_level'])
+            except ValueError as err:
+                self.warn('_set_level set to uknown level: {}'.format(err))
```
Comment (nir0s) [inline wryte.py:648]: uknown -> unknown

File: wryte.py [hunk 7]
```diff
@@ -602,44 +609,86 @@ def log(self, level, message, *objects, **kwargs):
...
     def critical(self, message, *objects, **kwargs):
         if '_set_level' in kwargs:
-            self.set_level(kwargs['_set_level'])
-        obj = self._enrich(message, 'critical', objects, kwargs)
-        self.logger.critical(obj)
-
+            try:
+                self.set_level(kwargs['_set_level'])
+            except ValueError as err:
+                self.warn('_set_level set to uknown level: {}'.format(err))
```
Comment (nir0s) [inline wryte.py:657]: uknown -> unknown

File: wryte.py [hunk 8]
```diff
@@ -549,20 +464,112 @@ def add_elasticsearch_handler(self):
...
+
+def _proxy_method(name):
+    def method(self, *args, **kwargs):
+        return getattr(self._log_engine, name)(*args, **kwargs)
+
+    method.__name__ = name
+    return method
+
+
+class WryteLogger(object):
```
Comment (nir0s) [inline wryte.py:482]: `(object)` is not required in Python3 (which we lint by).

File: wryte.py [hunk 9]
```diff
@@ -549,20 +464,112 @@ def add_elasticsearch_handler(self):
...
+    def __init__(self, log_engine, log_context=None):
+        self._log_engine = log_engine
+        if log_context:
+            self._log_context = log_context
+        else:
+            self._log_context = {}
+
```
Comment (nir0s) [inline wryte.py:489]: This upper part could be instead:
```python
self._log_context = log_engine._log # though this should no longer be private in the engine.
self._log_context.update(log_context or {})
```

File: wryte.py [hunk 10]
```diff
@@ -602,44 +609,86 @@ def log(self, level, message, *objects, **kwargs):
...
+            try:
+                self.set_level(kwargs['_set_level'])
+            except ValueError as err:
+                self.warn('_set_level set to uknown level: {}'.format(err))
+        obj = self._enrich(message, 'CRITICAL', objects, kwargs)
+        self._log_engine.logger.critical(obj)
+
+    set_level = _proxy_method('set_level')
+    get_level = _proxy_method('get_level')
+
```
Comment (nir0s) [inline wryte.py:663]: Oh, these are both just stuck here in the middle. I'd move them to the class definition.

**Comment by nir0s** [comment]
There are also many linting issues:
```
************* Module wryte
wryte.py:65:0: C0301: Line too long (139/120) (line-too-long)
wryte.py:77:0: C0330: Wrong hanging indentation.
            }
    |   |   ^ (bad-continuation)
wryte.py:288:43: C0303: Trailing whitespace (trailing-whitespace)
wryte.py:526:0: C0303: Trailing whitespace (trailing-whitespace)
wryte.py:663:0: C0303: Trailing whitespace (trailing-whitespace)
wryte.py:684:0: C0330: Wrong continued indentation (add 7 spaces).
                                    level=level, pretty=pretty,
                                    ^      | (bad-continuation)
wryte.py:685:0: C0330: Wrong continued indentation (add 7 spaces).
                                    bare=bare, jsonify=jsonify,
                                    ^      | (bad-continuation)
wryte.py:686:0: C0330: Wrong continued indentation (add 7 spaces).
                                    color=color,
                                    ^      | (bad-continuation)
wryte.py:687:0: C0330: Wrong continued indentation (add 7 spaces).
                                    simple=simple, enable_ec2=enable_ec2))
                                    ^      | (bad-continuation)
wryte.py:688:0: C0303: Trailing whitespace (trailing-whitespace)
wryte.py:63:23: W0212: Access to a protected member _nameToLevel of a client class (protected-access)
wryte.py:65:81: E1101: Module 'logging' has no '_levelNames' member (no-member)
wryte.py:65:81: W0212: Access to a protected member _levelNames of a client class (protected-access)
wryte.py:153:0: R0205: Class 'WryteLoggerEngine' inherits from object, can be safely removed from bases in python3 (useless-object-inheritance)
wryte.py:476:23: W0212: Access to a protected member _log_engine of a client class (protected-access)
wryte.py:482:0: R0205: Class 'WryteLogger' inherits from object, can be safely removed from bases in python3 (useless-object-inheritance)
```

File: wryte.py [hunk 11]
```diff
@@ -602,44 +609,86 @@ def log(self, level, message, *objects, **kwargs):
...
+                 name=None,
+                 hostname=None,
+                 level='INFO',
+                 pretty=None,
+                 bare=False,
+                 jsonify=False,
+                 color=True,
+                 simple=False,
+                 enable_ec2=False):
+        super().__init__(WryteLoggerEngine(name=name, hostname=hostname,
```
Comment (nir0s) [inline wryte.py:683]: Won't work on py2.7. `super` must be passed with the class.

File: wryte.py [hunk 12]
```diff
@@ -549,20 +464,112 @@ def add_elasticsearch_handler(self):
...
+        return getattr(self._log_engine, name)(*args, **kwargs)
+
+    method.__name__ = name
+    return method
+
+
+class WryteLogger(object):
+    def __init__(self, log_engine, log_context=None):
+        self._log_engine = log_engine
+        if log_context:
```
Comment (nir0s) [inline wryte.py:485]: This won't even work with the previously set context. You're referring to `log_context` as the only context there is, but there is also implicit context generated by `_get_base` in the engine that is exposed in `self.log_engine._log`.

This, for instance, is preventing the `ConsoleFormatter` from working at all.

File: wryte.py [hunk 13]
```diff
@@ -549,20 +464,112 @@ def add_elasticsearch_handler(self):
...
-            return
-
         self.logger.setLevel(level.upper())
 
+    def get_level(self):
+        return self.logger.getEffectiveLevel()
+
+
+
+def _proxy_method(name):
```
Comment (nir0s) [inline wryte.py:474]: This requires a serious explanation in-code. It's clear why this is implemented only on deeper reading of the code.

File: wryte.py [hunk 14]
```diff
@@ -549,20 +464,112 @@ def add_elasticsearch_handler(self):
     def set_level(self, level):
         """Set the current logger instance's level.
         """
-        if not self._assert_level(level):
-            return
-
         self.logger.setLevel(level.upper())
 
+    def get_level(self):
```
Comment (nir0s) [inline wryte.py:469]: Great addition

Comment (nir0s) [reply wryte.py:469]: I'd add a docstring for consistency.

File: wryte.py [hunk 15]
```diff
@@ -549,20 +464,112 @@ def add_elasticsearch_handler(self):
...
+
+        return log
+
     def bind(self, *objects, **kwargs):
         """Bind context to the logger's instance.
 
         After binding, each log entry will contain the bound fields.
         """
-        self._log.update(self._normalize_objects(objects))
+        self._log_context.update(self._normalize_objects(objects))
```
Comment (nir0s) [inline wryte.py:569]: I like the name. Makes much more sense than `_log`.

File: wryte.py [hunk 16]
```diff
@@ -602,44 +609,86 @@ def log(self, level, message, *objects, **kwargs):
...
+                self.set_level(kwargs['_set_level'])
+            except ValueError as err:
+                self.warn('_set_level set to uknown level: {}'.format(err))
+        obj = self._enrich(message, 'CRITICAL', objects, kwargs)
+        self._log_engine.logger.critical(obj)
+
+    set_level = _proxy_method('set_level')
+    get_level = _proxy_method('get_level')
+    
+    def nested(self, context=None, **kwargs):
```
Comment (nir0s) [inline wryte.py:664]: Please add docstring.

File: wryte.py [hunk 17]
```diff
@@ -602,44 +609,86 @@ def log(self, level, message, *objects, **kwargs):
...
+        _context = copy.deepcopy(self._log_context)
+        if isinstance(context, dict):
+            _context.update(context)
```
Comment (nir0s) [inline wryte.py:667]: What if `context` isn't a dict. You thik we should let it slide?

File: wryte.py (hunk 9 above)
Comment (nir0s) [reply wryte.py:489]: Honestly, I'm not sure I'd add the previous context implicitly. Maybe it should be sent as another argument.

**Review by nir0s — CHANGES_REQUESTED** [review]

**Comment by nir0s** [comment]
FYI @wolf-daniel

**Comment by m1keil** [comment]
going to close it for now :)

**Pull request closed without merging**